### PR TITLE
GEM-229 | Asynchronous settlements

### PR DESCRIPTION
**Pull request by lawos98**

### CODE REVIEW

File: src/test/kotlin/pl/edu/agh/gem/internal/model/expense/AcceptedExpenseTest.kt
Comment (pwgacek) [inline src/test/kotlin/pl/edu/agh/gem/internal/model/expense/AcceptedExpenseTest.kt:1]: please change BalanceElements to BalanceList in test names

File: src/main/kotlin/pl/edu/agh/gem/internal/model/expense/AcceptedExpense.kt
Comment (pwgacek) [inline src/main/kotlin/pl/edu/agh/gem/internal/model/expense/AcceptedExpense.kt:1]: i think that we should revert sign (if A pays of B and C 4$ each Balance shouldBe: A:-8$, B:4$, C:4$, but currently it is A:8$, B:-4$, C:-4$) Same in AcceptedPayments

File: src/integration/kotlin/pl/edu/agh/gem/integration/BaseIntegrationSpec.kt [hunk 1]
```diff
@@ -5,15 +5,35 @@ import mu.KotlinLogging
...
+        @JvmStatic
+        @DynamicPropertySource
+        fun injectContainerData(registry: DynamicPropertyRegistry) {
+            logger.info { "Injecting configuration" }
+            ProjectConfig.updateConfiguration(registry)
+        }
+
+        val testClock = Clock.systemUTC()
+        val FIXED_TIME = Instant.parse("2021-01-01T00:00:00Z")
+        val FIXED_WEDNESDAY = Instant.parse("2021-01-06T00:00:00Z")
```
Comment (pwgacek) [inline src/integration/kotlin/pl/edu/agh/gem/integration/BaseIntegrationSpec.kt]: not used

File: src/main/kotlin/pl/edu/agh/gem/internal/service/FinanceService.kt [hunk 2]
```diff
@@ -31,30 +36,39 @@ class FinanceService(
...
-    fun getBalances(groupId: String): Balances {
-        val balances = buildBalancesMap(groupId)
+    fun getBalances(groupId: String): List<Balances> {
+        return balancesRepository.getBalances(groupId)
+    }
 
-        val expenseBalanceElements = expenseManagerClient.getAcceptedExpenses(groupId).flatMap { it.toBalanceElements() }
-        val paymentBalanceElements = paymentManagerClient.getAcceptedPayments(groupId).flatMap { it.toBalanceElements() }
+    fun fetchBalances(groupId: String, currency: String): Balances {
+        val expenseBalanceList = expenseManagerClient.getAcceptedExpenses(groupId)
```
Comment (pwgacek) [inline src/main/kotlin/pl/edu/agh/gem/internal/service/FinanceService.kt]: please modify it to use updated versions of expense and payment managers

File: src/main/kotlin/pl/edu/agh/gem/external/persistence/balance/SettlementsEntity.kt
Comment (pwgacek) [inline src/main/kotlin/pl/edu/agh/gem/external/persistence/balance/SettlementsEntity.kt:1]: please rename file

File: src/main/kotlin/pl/edu/agh/gem/internal/job/ProcessingStage.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    fun nextStage(reconciliationJob: ReconciliationJob, nextState: ReconciliationJobState): StageResult {
+        return NextStage(reconciliationJob, nextState)
+    }
+
+    fun success(): StageResult {
+        return StageSuccess
+    }
+
+    fun failure(exception: Exception): StageResult {
```
Comment (pwgacek) [inline src/main/kotlin/pl/edu/agh/gem/internal/job/ProcessingStage.kt:18]: should it be not used?

**Review by pwgacek — CHANGES_REQUESTED** [review]

File: src/main/kotlin/pl/edu/agh/gem/internal/service/FinanceService.kt [hunk 4]
```diff
@@ -31,30 +36,39 @@ class FinanceService(
...
-            memberIds.associateWith { BigDecimal.ZERO }.toMutableMap()
-        }.toMutableMap()
+    fun saveBalances(balances: Balances) {
+        balancesRepository.save(balances)
     }
 
-    fun getBalances(groupId: String): Balances {
-        val balances = buildBalancesMap(groupId)
+    fun getBalances(groupId: String): List<Balances> {
+        return balancesRepository.getBalances(groupId)
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/internal/service/FinanceService.kt]: check if we have all users and all currenceis in the balances list if not we fetch new

File: src/main/kotlin/pl/edu/agh/gem/internal/model/expense/AcceptedExpense.kt
Comment (lawos98) [reply src/main/kotlin/pl/edu/agh/gem/internal/model/expense/AcceptedExpense.kt:1]: no i think is A pays for that expense its balance should be grater than 0 so if all participants will make settlements, A will be +8 not -8

Comment (pwgacek) [reply src/main/kotlin/pl/edu/agh/gem/internal/model/expense/AcceptedExpense.kt:1]: Settlements are made from positive balances (B :4€, C:4€) to negative (-8€) thats how work all allgorithms. So making it reveresd is not intuitive

Comment (lawos98) [reply src/main/kotlin/pl/edu/agh/gem/internal/model/expense/AcceptedExpense.kt:1]: Settlements yes, but we talk about balances

Comment (pwgacek) [reply src/main/kotlin/pl/edu/agh/gem/internal/model/expense/AcceptedExpense.kt:1]: ok

**Review by pwgacek — APPROVED** [review]

**Pull request merged**